base/iterators.jl

```diff
@@ -1095,4 +1094,37 @@ eltype(::Type{Stateful{T, VS}} where VS) where {T} = eltype(T)
 IteratorEltype(::Type{Stateful{T,VS}}) where {T,VS} = IteratorEltype(T)
 length(s::Stateful) = length(s.itr) - s.taken
 
+"""
+    only(x)
+
+Returns the one and only element of collection `x`, and throws an `ArgumentError` if the
+collection has zero or multiple elements.
+"""
+@propagate_inbounds function only(x)
+    i = iterate(x)
+    @boundscheck if i === nothing
+        throw(ArgumentError("Collection is empty, must contain exactly 1 element"))
+    end
+    (ret, state) = i
+    @boundscheck if iterate(x, state) !== nothing
+        throw(ArgumentError("Collection has multiple elements, must contain exactly 1 element"))
+    end
+    return ret
+end
+
+# Collections of known size
+only(x::Ref) = x[]
+only(x::Number) = x
+only(x::Char) = x
+only(x::Tuple{}) = throw(ArgumentError("Tuple is empty, must contain exactly 1 element"))
```

Comment:
What's the reason for adding special-cased extra methods that just throws? Seems like they can all just go to the generic `only(::Any)` method above?

Reply:
I guess same as all the other special-casing i.e. why do extra work?
```julia
julia> @code_typed only(())
CodeInfo(
1 ── %1  = Base.sle_int(1, 1)::Bool
└───       goto #3 if not %1
2 ── %3  = Base.sle_int(1, 0)::Bool
└───       goto #4
3 ──       nothing::Nothing
4 ┄─ %6  = φ (#2 => %3, #3 => false)::Bool
└───       goto #6 if not %6
5 ──       invoke Base.getindex(_2::Tuple, 1::Int64)::Union{}
└───       $(Expr(:unreachable))::Union{}
6 ┄─ %10 = Base.nothing::Nothing
└───       goto #7
7 ──       goto #8
8 ──       goto #11 if not $(Expr(:boundscheck))
9 ──       goto #11 if not true
10 ─ %15 = %new(Core.ArgumentError, "Collection is empty, must contain exactly 1 element")::ArgumentError
│          Main.throw(%15)::Union{}
└───       $(Expr(:unreachable))::Union{}
11 ┄       invoke Base.indexed_iterate(%10::Nothing, 1::Int64)::Union{}
└───       $(Expr(:unreachable))::Union{}
) => Union{}

julia> @code_warntype only(())
Variables
  #self#::Core.Compiler.Const(only, false)
  x::Core.Compiler.Const((), false)
  i::Nothing
  ret::Union{}
  @_5::Union{}
  state::Union{}

Body::Union{}
1 ─      nothing
│        Core.NewvarNode(:(ret))
│        Core.NewvarNode(:(@_5))
│        Core.NewvarNode(:(state))
│        (i = Main.iterate(x))
└──      goto #4 if not $(Expr(:boundscheck))
2 ─ %7 = (i === Main.nothing)::Core.Compiler.Const(true, false)
└──      goto #4 if not %7
3 ─ %9 = Main.ArgumentError("Collection is empty, must contain exactly 1 element")::ArgumentError
└──      Main.throw(%9)
4 ┄      Base.indexed_iterate(i, 1)
│        Core.Compiler.Const(:(ret = Core.getfield(%11, 1)), false)
│        Core.Compiler.Const(:(@_5 = Core.getfield(%11, 2)), false)
│        Core.Compiler.Const(:(Base.indexed_iterate(i, 2, @_5)), false)
│        Core.Compiler.Const(:(state = Core.getfield(%14, 1)), false)
│        $(Expr(:boundscheck))
│        Core.Compiler.Const(:(Main.iterate(x, state)), false)
│        Core.Compiler.Const(:(%17 !== Main.nothing), false)
│        Core.Compiler.Const(:(%18), false)
│        Core.Compiler.Const(:(Main.ArgumentError("Collection has multiple elements, must contain exactly 1 element")), false)
│        Core.Compiler.Const(:(Main.throw(%20)), false)
└──      Core.Compiler.Const(:(return ret), false)
```
```julia
julia> only(x::Tuple{}) = throw(ArgumentError("Tuple is empty, must contain exactly 1 element"))
only (generic function with 2 methods)

julia> @code_typed only(())
CodeInfo(
1 ─ %1 = %new(Core.ArgumentError, "Tuple is empty, must contain exactly 1 element")::ArgumentError
│        Main.throw(%1)::Union{}
└──      $(Expr(:unreachable))::Union{}
) => Union{}

julia> @code_warntype only(())
Variables
  #self#::Core.Compiler.Const(only, false)
  x::Core.Compiler.Const((), false)

Body::Union{}
1 ─ %1 = Main.ArgumentError("Tuple is empty, must contain exactly 1 element")::ArgumentError
│        Main.throw(%1)
└──      Core.Compiler.Const(:(return %2), false)
```

Reply:
Why does that matter when the end result is an error?

Reply:
Ah, yeah, i misspoke - sorry

Special casing errors in the cases of known length let's us give nicer error messages. 

I can see the case the the "usual" error would be fine for the `Tuple` case, 
but it is more user-friendly to include the lengths in the `Tuple` and `NamedTuple` cases

Reply:
Is it worth the extra code and extra methods just so that the user see `Tuple` instead of `Collection`?

Reply:
For that one method, I think not -- will ditch it

For the others that give the length, I do think it's pretty useful info to be able to include!